src/include/platform/internal/GenericConfigurationManagerImpl.ipp

```diff
@@ -52,6 +53,114 @@ namespace chip {
 namespace DeviceLayer {
 namespace Internal {
 
+#if CHIP_USE_TRANSITIONAL_DEVICE_INSTANCE_INFO_PROVIDER
+template <class ConfigClass>
+class LegacyDeviceInstanceInfoProvider : public DeviceInstanceInfoProvider
+{
+
+public:
+    // GenericConfigurationManagerImpl will own a LegacyDeviceInstanceInfoProvider which
+    // *refers back to that GenericConfigurationManagerImpl*, due to how CRTP-based
+    // storage APIs are defined. This is a bit unclean, but only applicable to the
+    // transition path when `CHIP_USE_TRANSITIONAL_DEVICE_INSTANCE_INFO_PROVIDER` is true.
+    // This circular dependency is NOT needed by DeviceInstanceInfoProvider, but required
+    // to keep legacy code running.
+    LegacyDeviceInstanceInfoProvider(GenericConfigurationManagerImpl<ConfigClass> & configManager) :
+        mGenericConfigManager(configManager)
+    {}
+
+    CHIP_ERROR GetSerialNumber(MutableCharSpan & aSerialNumber) override;
+    CHIP_ERROR GetManufacturingDate(MutableCharSpan & aManufacturingDate) override;
+    CHIP_ERROR GetHardwareVersion(uint16_t & aHardwareVersion) override;
+    CHIP_ERROR GetHardwareVersionString(MutableCharSpan & aHardwareVersionString) override;
+    CHIP_ERROR GetRotatingDeviceIdUniqueId(MutableByteSpan & aUniqueIdSpan) override;
+
+private:
+    GenericConfigurationManagerImpl<ConfigClass> & mGenericConfigManager;
+};
+
+template <class ConfigClass>
+CHIP_ERROR LegacyDeviceInstanceInfoProvider<ConfigClass>::GetSerialNumber(MutableCharSpan & aSerialNumber)
+{
+    ChipError err                                             = CHIP_NO_ERROR;
```

Comment:
I think you can simply move the implementation from `GenericConfigurationManagerImpl::GetSerialNumber`. In particular, make sure that CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER and other defines are used. Also, the moved functions can be removed from ConfigurationManager.

Reply:
That also means that we don't need no more keys in ZephyrConfig.

Reply:
Would we like to use `char*` (as in `GenericConfigurationManagerImpl::GetSerialNumber`) instead of MutableCharSpan?

Reply:
No, changing the type is fine, but if you keep to implementation close to the original, you will reduce chances of breaking other platforms/apps.

Reply:
Ok, I understand your intention. :)